### PR TITLE
Show `--help` if there is no man page for subcommand

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

part of https://github.com/rust-lang/cargo/issues/11408

- Show `--help` if there is no man page for the subcommand
- Do not suggest `cargo help config` until we add a man page.

### How should we test and review this PR?
1. build cargo
2. try `cargo help config`
3. see:
```sh
➜  cargo git:(master) ✗ ./target/debug/cargo help config
Inspect configuration values

Usage: cargo config [OPTIONS] <COMMAND>

Commands:
  get  

Options:
  -v, --verbose...          Use verbose output (-vv very verbose/build.rs output)
      --color <WHEN>        Coloring: auto, always, never
      --frozen              Require Cargo.lock and cache are up to date
      --locked              Require Cargo.lock is up to date
      --offline             Run without accessing the network
      --config <KEY=VALUE>  Override a configuration value
  -Z <FLAG>                 Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
  -h, --help                Print help information
```

### Additional information
I am not sure I should add a test for it because I think we will add a man page for config.

### CODE REVIEW

File: src/bin/cargo/commands/help.rs [hunk 1]
```diff
@@ -60,13 +61,13 @@ fn try_help(config: &Config, subcommand: &str) -> CargoResult<bool> {
     if resolve_executable(Path::new("man")).is_ok() {
         let man = match extract_man(subcommand, "1") {
             Some(man) => man,
-            None => return Ok(false),
+            None => bail!("no man page for `{}`", subcommand),
```
Comment (hi-rustin) [inline src/bin/cargo/commands/help.rs:64]: I am not sure if this is a good message because maybe users don't know why we need a man page. I'll improve it later.

File: src/bin/cargo/commands/help.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 use crate::aliased_command;
 use crate::command_prelude::*;
+use anyhow::bail;
```
Comment (hi-rustin) [inline src/bin/cargo/commands/help.rs:3]: Do I need to sort the use statements? Because I saw we did it in other files.

File: src/bin/cargo/commands/help.rs (hunk 1 above)
Comment (epage) [reply src/bin/cargo/commands/help.rs:64]: Rather than an error, why not show `--help`, like external subcommands?

File: src/bin/cargo/main.rs [hunk 3]
```diff
@@ -189,9 +189,22 @@ fn execute_external_subcommand(config: &Config, cmd: &str, args: &[&OsStr]) -> C
...
+}
+
+fn execute_internal_subcommand(config: &Config, args: &[&OsStr]) -> CliResult {
+    execute_subcommand(config, None, args)
+}
 
+// This function is used to execute a subcommand. It is used to execute both
+// internal and external subcommands.
+// If `cmd_path` is `None`, then the subcommand is an internal subcommand.
+fn execute_subcommand(config: &Config, cmd_path: Option<&PathBuf>, args: &[&OsStr]) -> CliResult {
```
Comment (hi-rustin) [inline src/bin/cargo/main.rs:202]: I am not sure does this is a good API design. Could you please take a look?

File: src/bin/cargo/commands/help.rs (hunk 1 above)
Comment (hi-rustin) [reply src/bin/cargo/commands/help.rs]: Updated. Good idea!

File: src/bin/cargo/commands/help.rs (hunk 2 above)
Comment (hi-rustin) [reply src/bin/cargo/commands/help.rs]: Got it. Thanks!

File: src/bin/cargo/commands/config.rs [hunk 4]
```diff
@@ -4,7 +4,6 @@ use cargo::ops::cargo_config;
 pub fn cli() -> Command {
     subcommand("config")
         .about("Inspect configuration values")
-        .after_help("Run `cargo help config` for more detailed information.\n")
```
Comment (epage) [inline src/bin/cargo/commands/config.rs:7]: Could we split this out into its own commit so we can just revert this when needed?

File: src/bin/cargo/commands/help.rs [hunk 5]
```diff
@@ -22,11 +22,21 @@ pub fn exec(config: &mut Config, args: &ArgMatches) -> CliResult {
...
-                config,
-                subcommand,
-                &[OsStr::new(subcommand), OsStr::new("--help")],
-            )?;
+            match check_builtin(&subcommand) {
+                Some(s) => {
+                    crate::execute_internal_subcommand(
+                        config,
+                        &[OsStr::new(s), OsStr::new("--help")],
+                    )?;
```
Comment (epage) [inline src/bin/cargo/commands/help.rs:30]: Side note: a part of me would rather we do this without spawning but that would require calling into the top-level `clap::Command` with our own custom args while this ensures we have a unified experience.

File: src/bin/cargo/main.rs (hunk 3 above)
Comment (epage) [reply src/bin/cargo/main.rs:202]: Not the greatest but the scope of it is fairly small so its likely not worth us spending a lot of time iterating on the most ideal internal API.  We can always iterate more on this later.

File: src/bin/cargo/commands/config.rs (hunk 4 above)
Comment (hi-rustin) [reply src/bin/cargo/commands/config.rs:7]: Split! See https://github.com/rust-lang/cargo/pull/11473/commits/53fd81552299d0c0a04bc5bfe682c868262d15b0

File: src/bin/cargo/commands/help.rs (hunk 5 above)
Comment (hi-rustin) [reply src/bin/cargo/commands/help.rs:30]: Did you mean like this?
```diff
diff --git a/src/bin/cargo/commands/help.rs b/src/bin/cargo/commands/help.rs
index 2839b931e..78fb3db59 100644
--- a/src/bin/cargo/commands/help.rs
+++ b/src/bin/cargo/commands/help.rs
@@ -22,12 +22,11 @@ pub fn exec(config: &mut Config, args: &ArgMatches) -> CliResult {
     let subcommand = args.get_one::<String>("COMMAND");
     if let Some(subcommand) = subcommand {
         if !try_help(config, subcommand)? {
-            match check_builtin(&subcommand) {
+            match super::builtin_exec(&subcommand) {
                 Some(s) => {
-                    crate::execute_internal_subcommand(
-                        config,
-                        &[OsStr::new(s), OsStr::new("--help")],
-                    )?;
+                    let cmd = crate::cli::cli();
+                    let matches = cmd.get_matches_from(["cargo", subcommand, "--help"]);
+                    s(config, &matches)?;
                 }
                 None => {
                     crate::execute_external_subcommand(

```

**Comment by hi-rustin** [comment]
Thanks for your review!  💚 💙 💜 💛 ❤️

File: src/bin/cargo/commands/help.rs (hunk 5 above)
Comment (epage) [reply src/bin/cargo/commands/help.rs:30]: Something like that, yeah.
